### PR TITLE
python3Packages.tlsprofiler: init at 1.0

### PR DESCRIPTION
**Pull request by veehaitch**
###### Motivation for this change

Tlsprofiler allows to test if a TLS server adheres to [Mozzilla's server side TLS recommendations](https://wiki.mozilla.org/Security/Server_Side_TLS#Recommended_configurations). NixOS also relies on these guidelines for [Nginx](https://github.com/NixOS/nixpkgs/blob/release-20.03/nixos/modules/services/web-servers/nginx/default.nix#L89-L103), implementing the "intermediate" profile as a configuration flag. The original authors of Tlsprofiler provide a web version [here](https://tlsprofiler.danielfett.de/).

Tlsprofiler makes use of Nassl/SSLyze, which have been merged [recently](https://github.com/NixOS/nixpkgs/pull/89034). It does, however, rely on a forked version of [Nassl](https://github.com/fabian-hk/nassl/tree/tls_profiler) and [SSLyze](https://github.com/fabian-hk/sslyze/tree/tls_profiler). In contrast to the Nixpgks version of SSLyze, I had to disable the tests as virtually all are online. They passed just fine though: https://gist.github.com/veehaitch/a7eb5164ff1480dd47da3500509897d3

One may use Tlsprofiler as a Python 3 package or as a command line application; see `tlsprofiler -h` for further information.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/tlsprofiler/default.nix [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    rev = "c4a9cdcf951343ef6cf670df9351c197c6aaab80";
+    sha256 = "1ng9ba1w6x9x86cngxx9p4dfjzkf3nn0w4ibn1kmwnf2rgdl6clw";
+  };
+
+  patches = [ ./tlsprofiler-setup-requirements.patch ];
+
+  # Tests require Docker to set up web servers which serve a specific profile
+  doCheck = false;
+
+  propagatedBuildInputs = [ requests cryptography nasslTlsprofiler sslyzeTlsprofiler ];
```
[the remark this answers is not in the document]
Comment (veehaitch) [reply pkgs/development/python-modules/tlsprofiler/default.nix:59]: Thanks for your review. I agree with your assessment that pinning these versions should be avoided; particularly, as it isn't unlikely that someone who uses `tlsprofiler` in their Python project might also use `sslyze`. I'd propose the following strategy:
1. Ask the authors of TLS Profiler if they think it is possible to make some efforts to merge their fork of Nassl/SSLyze with upstream. To that end, I've created an issue: https://github.com/danielfett/tlsprofiler/issues/6. If this works out, we could include a modified version of this PR without further problems, I guess.
2. If this won't happen for any reason, I'll modify the PR to include the application only (similar to `awscli`).

File: pkgs/development/python-modules/tlsprofiler/default.nix [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      };
+
+      patches = [
+        # Apply upstream patch to update expired cert chain
+        ./nassl-0001-Fix-test.patch
+      ];
+    }
+  );
+  sslyzeTlsprofiler = (sslyze.override { nassl = nasslTlsprofiler; }).overrideAttrs (
+    oldAttrs: rec {
```
Comment (samuela) [inline pkgs/development/python-modules/tlsprofiler/default.nix:28]: Why is `rec` needed?

File: pkgs/development/python-modules/tlsprofiler/tlsprofiler-setup-requirements.patch [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+@@ -22,6 +22,7 @@ setup(
+     python_requires='>=3.6',
+     install_requires=[
+         'requests',
+-        'sslyze @ git+https://github.com/fabian-hk/sslyze.git@tls_profiler#egg=sslyze',
++        'sslyze',
++        'cryptography',
+     ],
+ )
+
```
Comment (samuela) [inline pkgs/development/python-modules/tlsprofiler/tlsprofiler-setup-requirements.patch:23]: I think this patch would be easier with just two substituteInPlace's.

File: pkgs/development/python-modules/tlsprofiler/default.nix [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  patches = [ ./tlsprofiler-setup-requirements.patch ];
+
+  # Tests require Docker to set up web servers which serve a specific profile
+  doCheck = false;
+
+  propagatedBuildInputs = [ requests cryptography nasslTlsprofiler sslyzeTlsprofiler ];
+
+  # Also make `run.py` available as `tlsprofiler` application
+  postInstall = ''
+    sed -i '1s|^|#!/usr/bin/env python3\n|' run.py
```
Comment (samuela) [inline pkgs/development/python-modules/tlsprofiler/default.nix:63]: Why is this necessary?

File: pkgs/development/python-modules/tlsprofiler/default.nix [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      src = fetchFromGitHub {
+        owner = "fabian-hk";
+        repo = "nassl";
+        rev = "7321a149b98175ca20af41561eb04bcb3b7cba41";
+        sha256 = "124sx3b0a6bq1w0w5vx4prczmck8n46knxz19ma7acnw5mrdbnx4";
+      };
+
+      patches = [
+        # Apply upstream patch to update expired cert chain
+        ./nassl-0001-Fix-test.patch
```
Comment (samuela) [inline pkgs/development/python-modules/tlsprofiler/default.nix:23]: If this is an upstream patch can we use fetchpatch instead?

File: pkgs/development/python-modules/tlsprofiler/default.nix [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+, fetchFromGitHub
+, buildPythonPackage
+, nassl
+, sslyze
+, requests
+, cryptography
+}:
+let
+  nasslTlsprofiler = nassl.overrideAttrs (
+    oldAttrs: rec {
```
Comment (samuela) [inline pkgs/development/python-modules/tlsprofiler/default.nix:11]: I don't think `rec` is necessary.

**Review by samuela — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
Closing due to no response from author in a while.

**Pull request closed without merging**